keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,376 @@
+* RestartPod - restart the whole Pod in-place to apply new resource values
+  (e.g. Pod requires its Init Containers to re-run).
+
+By using the ResizePolicy flag, user can mark Containers or Pods as safe
+(or unsafe) for in-place resource update.
+
+This flag is used by Kubelet to determine the actions needed. This flag **may** be
+used by the actors starting the update to decide if the process should be started
+at all (for example VPA might decide to evict Pod with RestartPod policy).
+
+Setting the flag to separately control CPU & memory is due to an observation
+that usually CPU can be added/removed without much problem whereas changes to
+available memory are more probable to require restarts.
+
+If more than one resource type with different policies are updated, then
+RestartPod policy takes precedence over RestartContainer, which in turn takes
+precedence over NoRestart policy.
+
+#### CRI Changes
+
+Kubelet calls UpdateContainerResources CRI API which currently takes
+*runtimeapi.LinuxContainerResources* parameter that works for Docker and Kata,
+but not for Windows. This parameter is changed to *runtimeapi.ContainerResources*,
+that is runtime agnostic.
+
+### Flow Control
+
+When a new Pod is created, Scheduler is responsible for selecting a suitable
+Node that accommodates the Pod. When a Pod resize is requested, Kubelet attempts
+to update the Pod and its Containers. If required, Kubelet may evict lower
+priority Pods from the Node to make room for the resize. Scheduler observes the
+resize request, and updates the accounting in its cache.
+
+The following steps denote a typical flow of an in-place resize operation for a Pod
+with ResizePolicy set to NoRestart for all its Containers.
+
+1. Initiating actor updates Pod's Container.ResourceRequirements using PATCH verb.
+1. API Server validates the new ResourceRequirements (e.g. limits are not below
+   requested resources, QoS class does not change, ResourceQuota not exceeded..).
+1. API Server calls all Admission Controllers to verify the Pod Update.
+   * If any of the Controllers reject the update, API Server responds with an
+     appropriate error message.
+1. API Server updates PodSpec object with the new desired ResourceRequirements.
+1. Kubelet observes that Pod's Container.ResourceRequirements and
+   ContainerStatus.ResourcesAllocated differ. It checks its Node allocatable
+   resources to determine if the new desired ResourceRequirements fit the Node.
+   * _Case 1_: Kubelet finds new ResourceRequirements fit. It sets
+     ResizingPod.Status to true, and ResizingPod.Reason to InProgress, then
+     applies resized cgroup limits to the Pod and its Containers, and once
+     successfully done, updates Pod's ContainerStatus.ResourcesAllocated to
+     reflect the desired ResourceRequirements. It then sets ResizingPod.Status
+     to false, and ResizingPod.Reason to Success.
+     - If at the same time, a new Pod was scheduled against the capacity used by
```

Comment:
i assume we could handle this in both size up and down scenarios?  what is the kubelet verifying during its admission pass?  is it looking at a sum of resource allocated across all pods relative to node allocatable?

Reply:
During HandlePodUpdate, we check the Pod lifecycle predicate Admit (which sums up resources of all Pods on the node) to see if the new desired resources delta will fit.  For size down (negative delta), this predicate should always succeed. For size up, it may fail due to race condition.

Reply:
Resolving the issue, please reopen if above response does not address the concerns.